### PR TITLE
Implement new-style resource state handling for GKE cluster

### PR DESCRIPTION
**Pull request by rileykarson**
Implements go/tpg-resource-state-handling, pre-work for https://github.com/terraform-providers/terraform-provider-google/issues/3304

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
container: `google_container_cluster` will allow importing / updating / deleting clusters in error states and will wait for a stable state after updates.
```

### CODE REVIEW

File: third_party/terraform/resources/resource_container_cluster.go.erb [hunk 1]
```diff
@@ -2113,22 +2123,38 @@ func cleanFailedContainerCluster(d *schema.ResourceData, meta interface{}) error
...
+var containerClusterRestingStates = RestingStates{
+	"RUNNING":  ReadyState,
+	"DEGRADED": ErrorState,
+	"ERROR":    ErrorState,
+}
+
+func containerClusterAwaitRestingState(config *Config, project, location, clusterName string, timeout time.Duration) (state string, stateType StateType, err error) {
+	err = resource.Retry(timeout, func() *resource.RetryError {
 		name := containerClusterFullName(project, location, clusterName)
 		cluster, err := config.clientContainerBeta.Projects.Locations.Clusters.Get(name).Do()
```
Comment (chrisst) [inline third_party/terraform/resources/resource_container_cluster.go.erb]: nit: local variables with the same name as named return variables is a warning flag for me. The scoping looks safer here but I'd suggest renaming 1 of them for clarity.

File: third_party/terraform/resources/resource_container_cluster.go.erb [hunk 2]
```diff
@@ -2113,22 +2123,38 @@ func cleanFailedContainerCluster(d *schema.ResourceData, meta interface{}) error
...
 		} else {
-			return resource.NonRetryableError(fmt.Errorf("Cluster %q has terminal state %q with message %q.", clusterName, cluster.Status, cluster.StatusMessage))
+			return resource.RetryableError(fmt.Errorf("Cluster %q has status %q with message %q", clusterName, state, cluster.StatusMessage))
 		}
 	})
+
+	if _, ok := containerClusterRestingStates[state]; ok {
+		return state, stateType, nil
+	} else {
+		return "", UndefinedState, err
```
Comment (chrisst) [inline third_party/terraform/resources/resource_container_cluster.go.erb]: you can simplify to 1 return here as `containerClusterRestingStates["garbage"] == UndefinedState`

File: third_party/terraform/resources/resource_container_cluster.go.erb [hunk 3]
```diff
@@ -2012,6 +2018,10 @@ func resourceContainerClusterUpdate(d *schema.ResourceData, meta interface{}) er
 <% end -%>
 	d.Partial(false)
 
+	if _, _, err := containerClusterAwaitRestingState(config, project, location, clusterName, d.Timeout(schema.TimeoutDelete)); err != nil {
```
Comment (chrisst) [inline third_party/terraform/resources/resource_container_cluster.go.erb]: timeoutUpdate

File: third_party/terraform/resources/resource_container_cluster.go.erb [hunk 4]
```diff
@@ -1225,9 +1225,15 @@ func resourceContainerClusterCreate(d *schema.ResourceData, meta interface{}) er
 		return err
 	}
 
-	if err := waitForContainerClusterReady(config, project, location, clusterName, d.Timeout(schema.TimeoutCreate)); err != nil {
+	state, stateType, err := containerClusterAwaitRestingState(config, project, location, clusterName, d.Timeout(schema.TimeoutCreate))
```
[the remark this answers is not in the document]
Comment (rileykarson) [reply third_party/terraform/resources/resource_container_cluster.go.erb]: I'd returned `stateType` to save a check against `containerClusterAwaitRestingState` but that was silly. I'll just return `state`. In cases where we reached a resting state, we return a success through a nil error and the state as metadata, for the caller to process.

**Pull request merged**